src/snowflake/snowpark/_internal/data_source/datasource_partitioner.py

```diff
@@ -0,0 +1,229 @@
+        self.dialect = dialect_class()
+        self.driver = driver_class(create_connection)
+        self.driver_class = driver_class
+        self.dialect_class = dialect_class
+        self.create_connection = create_connection
+
+    def reader(self) -> DataSourceReader:
+        return DataSourceReader(self.driver_class, self.create_connection)
+
+    def schema(
+        self,
+        table_or_query: str,
+        custom_schema: Optional[Union[str, StructType]] = None,
+    ) -> StructType:
+        if custom_schema is None:
+            return self.driver.infer_schema_from_description(table_or_query)
+        else:
+            if isinstance(custom_schema, str):
+                schema = type_string_to_type_object(custom_schema)
+                if not isinstance(schema, StructType):
+                    logger.error(f"Invalid schema string: {custom_schema}")
+                    raise ValueError(
+                        f"Invalid schema string: {custom_schema}. "
+                        f"You should provide a valid schema string representing a struct type."
+                    )
+                return schema
+            elif isinstance(custom_schema, StructType):
+                return custom_schema
+            else:
+                logger.error(f"Invalid schema type: {type(custom_schema)}")
+                raise TypeError(f"Invalid schema type: {type(custom_schema)}.")
+
+    def partitions(
+        self,
+        table_or_query: str,
+        schema: StructType,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        predicates: Optional[List[str]] = None,
+    ) -> List[Any]:
+        select_query = self.dialect.generate_select_query(table_or_query, schema)
+        logger.debug(f"Generated select query: {select_query}")
+        if column is None:
+            if (
+                lower_bound is not None
+                or upper_bound is not None
+                or num_partitions is not None
+            ):
+                raise ValueError(
+                    "when column is not specified, lower_bound, upper_bound, num_partitions are expected to be None"
+                )
+            if predicates is None:
+                partitioned_queries = [select_query]
+            else:
+                partitioned_queries = self.generate_partition_with_predicates(
+                    select_query, predicates
+                )
+        else:
+            if lower_bound is None or upper_bound is None or num_partitions is None:
+                raise ValueError(
+                    "when column is specified, lower_bound, upper_bound, num_partitions must be specified"
+                )
+
+            column_type = None
+            for field in schema.fields:
+                if field.name.lower() == column.lower():
+                    column_type = field.datatype
```

Comment:
nit: you can break the loop after you find the column right?
```suggestion
            for field in schema.fields:
                if field.name.lower() == column.lower():
                    column_type = field.datatype
                    break
```